src/Workspaces/Core/Portable/Utilities/ReferenceHolder`1.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Roslyn.Utilities
+{
+    internal readonly struct ReferenceHolder<T> : IEquatable<ReferenceHolder<T>>
+        where T : class?
+    {
+        [AllowNull, MaybeNull]
+        private readonly T _strongReference;
+        private readonly WeakReference<T>? _weakReference;
+        private readonly int _hashCode;
+
+        private ReferenceHolder(T strongReference)
+        {
+            _strongReference = strongReference;
+            _weakReference = null;
+            _hashCode = 0;
+        }
+
+        private ReferenceHolder(WeakReference<T> weakReference, int hashCode)
+        {
+            _strongReference = null;
+            _weakReference = weakReference;
+            _hashCode = hashCode;
+        }
+
+        public static ReferenceHolder<T> Strong(T value)
+            => new ReferenceHolder<T>(value);
+
+        public static ReferenceHolder<T> Weak(T value)
+        {
+            if (value is null)
+            {
+                // Track this as a strong reference so we know 'Equals' should only look at values originally null.
+                return Strong(value);
+            }
+
+            return new ReferenceHolder<T>(new WeakReference<T>(value), EqualityComparer<T>.Default.GetHashCode(value));
+        }
+
+        [return: MaybeNull]
+        public T TryGetTarget()
+        {
+            if (_weakReference is object)
+                return _weakReference.GetTarget();
+
+            return _strongReference;
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is ReferenceHolder<T> other
+                && Equals(other);
+        }
+
+        public bool Equals(ReferenceHolder<T> other)
+        {
+            var x = TryGetTarget();
+            var y = other.TryGetTarget();
+            if (x is null)
+            {
+                // Null values are equal iff both were originally references to null.
+                return y is null && other._weakReference is null;
+            }
+
+            // Intentional reference equality check
+            return x == y;
```

Comment:
ok, if we're doing intentional ref equality, should we do intentional ref GetHashCode?

Reply:
No. `GetHashCode` tends to be faster and provides the same guarantees.

Reply:
I may end up switching to RuntimeHelpers.GetHashCode as part of addressing the reflexivity issue.

Reply:
Switch completed.